src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs

```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
         }
     }
 
+    private static TypeMapping? TryBuildImplicitMapping(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        if (
+            !elementMapping.IsSynthetic
+            || ctx.MapperConfiguration.UseDeepCloning
+            || ctx.CollectionInfos!.Target.CollectionType == CollectionType.None
+        )
+            return null;
+
+        if (
+            ctx.CollectionInfos.Source.IsArray
+            && ctx.CollectionInfos.Target.CollectionType
+                is (
+                    CollectionType.IEnumerable
+                    or CollectionType.IReadOnlyCollection
+                    or CollectionType.IReadOnlyList
+                    or CollectionType.Array
+                )
+        )
+        {
+            {
+                return new CastMapping(ctx.Source, ctx.Target);
+            }
+        }
+
+        if (
+            !ctx.CollectionInfos.Source.IsArray
+            && ctx.CollectionInfos.Source.ImplementedTypes.HasFlag(ctx.CollectionInfos.Target.CollectionType)
+        )
+        {
+            {
+                return new CastMapping(ctx.Source, ctx.Target);
+            }
+        }
+
+        return null;
+    }
+
+    private static TypeMapping? TryBuildFastConversion(MappingBuilderContext ctx, ITypeMapping elementMapping)
```

Comment:
IMO this method gets a lot easier to read with a few little restructurings:
* move the `if (ctx.CollectionInfos.Target.CollectionType is CollectionType.List or CollectionType.ICollection or CollectionType.IList)` part up (just after the early null return)
* inverse the `ctx.CollectionInfos.Target.CollectionType` check and return null early
* inverse the `ctx.CollectionInfos.Source.CollectionType == CollectionType.Array` condition and return the `ArrayForEachMapping` early